### PR TITLE
feat: add built-in metric constants

### PR DESCRIPTION
**Pull request by mutianf**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigtable/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

### CODE REVIEW

File: google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/BuiltinMeasureConstants.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ */
+package com.google.cloud.bigtable.stats;
+
+import static io.opencensus.stats.Measure.MeasureLong;
+
+import com.google.api.core.InternalApi;
+import io.opencensus.tags.TagKey;
+
+@InternalApi("For internal use only")
+public class BuiltinMeasureConstants {
```
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/BuiltinMeasureConstants.java:24]: Please add some javadoc explaining that this is just the internal write side and under what prefix the metrics can be read?

File: google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/StatsWrapper.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+@InternalApi("For internal use only")
+public class StatsWrapper {
+
+  private final StatsComponent component;
+  private final boolean useLocalRecorder;
+
+  public StatsWrapper(boolean useLocalRecorder) {
+    this.component = new StatsComponentImpl();
+    this.useLocalRecorder = useLocalRecorder;
+  }
```
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/StatsWrapper.java:45]: I think you can make this a bit nicer:

```
public static StatsWrapper get() {
   return new StatsWrapper(Stats. getViewManager(), Stats. getStatsRecorder());
}

@InternalApi("VisibleForTesting")
public static StatsWrapper createPrivateInstance() {
    c= new StatsComponentImpl();
    return new StatsWrapper(c.getViewManager() ...)
}


private StatsWrapper(ViewManager, StatsRecorder) {...
```

File: google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/StatsWrapper.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import io.opencensus.stats.ViewData;
+import io.opencensus.stats.ViewManager;
+import io.opencensus.tags.TagKey;
+import io.opencensus.tags.TagValue;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/** Wrapper class for accessing opencensus * */
```
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/StatsWrapper.java:35]: Add a bit more doc here explaining that the builtin metrics use a shaded version of opencensus to avoid poluting the global opencensus namespace, so this provides a facade that will not be relocated

File: google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/StatsWrapper.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+  private final StatsComponent component;
+  private final boolean useLocalRecorder;
+
+  public StatsWrapper(boolean useLocalRecorder) {
+    this.component = new StatsComponentImpl();
+    this.useLocalRecorder = useLocalRecorder;
+  }
+
+  StatsRecorder getStatsRecorder() {
```
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/StatsWrapper.java:47]: why not make this public and return a BuiltinMetricsRecorder (similar to how stats owns StatsRecorder)?

File: google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/StatsWrapper.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+  ViewManager getViewManager() {
+    if (!useLocalRecorder) {
+      return Stats.getViewManager();
+    } else {
+      return component.getViewManager();
+    }
+  }
+
+  long getAggregationValueAsLong(
```
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/StatsWrapper.java:63]: Should this be here? it seems like it can be a helper that lives in tests?

File: google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/BuiltinMeasureConstants.java (hunk 1 above)
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/BuiltinMeasureConstants.java:24]: Should this be public?

**Review by igorbernstein2 — COMMENTED** [review]
Also, please add a README explaining what this artifact is

File: google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/BuiltinMetricsRecorder.java [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+  private final Tagger tagger;
+  private final StatsRecorder statsRecorder;
+  private final TagContext parentContext;
+  private final SpanName spanName;
+  private final Map<String, String> statsAttributes;
+
+  private MeasureMap attemptLevelNoStreaming;
+  private MeasureMap attemptLevelWithStreaming;
+  private MeasureMap operationLevelNoStreaming;
+  private MeasureMap operationLevelWithStreaming;
```
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/BuiltinMetricsRecorder.java:46]: the WithStreamin and NoStreaming is very hard to follow. Do non-streaming operations not have latencies?

File: google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/BuiltinMetricsRecorder.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+import io.opencensus.tags.TagContextBuilder;
+import io.opencensus.tags.TagKey;
+import io.opencensus.tags.TagValue;
+import io.opencensus.tags.Tagger;
+import io.opencensus.tags.Tags;
+import java.util.Map;
+
+/** Add built-in metrics to the measure map * */
+@InternalApi("For internal use only")
+public class BuiltinMetricsRecorder {
```
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/BuiltinMetricsRecorder.java:33]: why not stayt consistent call it StatsRecorderWrapper?

File: google-cloud-bigtable-stats/src/test/java/com/google/cloud/bigtable/stats/StatsRecorderWrapperTest.java [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+  @Before
+  public void setup() {
+    this.wrapper = StatsWrapper.createPrivateInstance();
+    BuiltinViews views = new BuiltinViews(wrapper);
+    views.registerBigtableBuiltinViews();
+  }
+
+  @Test
+  public void testStreamingOperation() throws InterruptedException {
+    StatsRecorderWrapper tracer =
```
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/src/test/java/com/google/cloud/bigtable/stats/StatsRecorderWrapperTest.java:57]: nit, not a tracer

**Review by igorbernstein2 — APPROVED** [review]
lgtm after nits

File: google-cloud-bigtable-stats/pom.xml [hunk 9]
```diff
@@ -38,8 +47,30 @@
...
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.grpc</groupId>
+            <artifactId>grpc-api</artifactId>
+            <scope>test</scope>
```
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/pom.xml:73]: nit for shade plugin below: now that this isnt just a packaging artifact, please fix the dependency plugin config below

File: google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/BuiltinMeasureConstants.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+class BuiltinMeasureConstants {
+  // TagKeys
+  static final TagKey PROJECT_ID = TagKey.create("project_id");
+  static final TagKey INSTANCE_ID = TagKey.create("instance_id");
+  static final TagKey APP_PROFILE = TagKey.create("app_profile");
+  static final TagKey METHOD = TagKey.create("method");
+  static final TagKey STREAMING = TagKey.create("streaming");
+  static final TagKey STATUS = TagKey.create("status");
+  static final TagKey CLIENT_NAME = TagKey.create("client_name");
+  static final TagKey CLIENT_ID = TagKey.create("client_id");
```
[the remark this answers is not in the document]
Comment (mutianf) [reply google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/BuiltinMeasureConstants.java:32]: This will be generated in the exporter. So creating the placeholder for now

File: google-cloud-bigtable-bom/pom.xml [hunk 11]
```diff
@@ -94,6 +94,11 @@
           <artifactId>proto-google-cloud-bigtable-v2</artifactId>
           <version>2.8.1-SNAPSHOT</version><!-- {x-version-update:proto-google-cloud-bigtable-v2:current} -->
         </dependency>
+          <dependency>
+              <groupId>com.google.cloud</groupId>
+              <artifactId>google-cloud-bigtable-stats</artifactId>
+              <version>2.8.1-SNAPSHOT</version><!-- {x-version-update:proto-google-cloud-bigtable-v2:current} -->
```
Comment (igorbernstein2) [inline google-cloud-bigtable-bom/pom.xml:100]: I think you want {x-version-update:google-cloud-bigtable:current} or to create your own

File: google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/StatsRecorderWrapper.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    TagContextBuilder tagCtx =
+        newTagContextBuilder(tableId, zone, cluster)
+            .putLocal(BuiltinMeasureConstants.STATUS, TagValue.create(status));
+
+    if (operationType == OperationType.ServerStreaming
+        && spanName.getMethodName().equals("ReadRows")) {
+      tagCtx.putLocal(BuiltinMeasureConstants.STREAMING, TagValue.create("true"));
+    } else {
+      tagCtx.putLocal(BuiltinMeasureConstants.STREAMING, TagValue.create("false"));
+    }
```
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/StatsRecorderWrapper.java:70]:
```
boolean isStreaming = operationType == OperationType.ServerStreaming;
tagCtx.putLocal(BuiltinMeasureConstants.STREAMING, TagValue.create(Boolean.toString(isStreaming)));
```

File: google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/StatsRecorderWrapper.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+  private final SpanName spanName;
+  private final Map<String, String> statsAttributes;
+
+  private MeasureMap measureMap;
+
+  public StatsRecorderWrapper(
+      OperationType operationType,
+      SpanName spanName,
+      Map<String, String> statsAttributes,
+      StatsWrapper statsWrapper) {
```
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/src/main/java/com/google/cloud/bigtable/stats/StatsRecorderWrapper.java:49]: I think you should make StatsWrapper be the factory for StatsRecorderWrapper

```java
class StatsWrapper {
  StatsRecorderWrapper createRecorder(OperationType operationType,
      SpanName spanName,
      Map<String, String> statsAttributes) {
     return new StatsRecorderWrapper(operationType, spanName, statsAttributes, statsRecorder, tagger);
 }
}

StatsRecorderWrapper(OperationType operationType,
      SpanName spanName,
      Map<String, String> statsAttributes,
     StatsRecorder recorder,
     Tagger tagger) 
```

This will better decouple the 2 and make it so that the child doesnt have to know about the parrent and be tested easier

File: google-cloud-bigtable-stats/src/test/java/com/google/cloud/bigtable/stats/StatsRecorderWrapperTest.java [hunk 14]
```diff
@@ -0,0 +1,511 @@
...
+            return arg.getCount();
+          }
+        },
+        new io.opencensus.common.Function<AggregationData.DistributionData, Long>() {
+          @Override
+          public Long apply(AggregationData.DistributionData arg) {
+            return (long) arg.getMean();
+          }
+        },
+        new io.opencensus.common.Function<AggregationData.LastValueDataDouble, Long>() {
```
Comment (igorbernstein2) [inline google-cloud-bigtable-stats/src/test/java/com/google/cloud/bigtable/stats/StatsRecorderWrapperTest.java:492]: why the full class name? why not an import?
Better yet, can you just use lambdas here?

**Review by igorbernstein2 — APPROVED** [review]
lgtm!
